File: src/primo_mcp_server/citations.py
"""Citation formatting for Primo records.

Supports APA 7th, Harvard, Chicago, IEEE, and Vancouver styles.
"""

from __future__ import annotations

from primo_mcp_server.models import PrimoRecord


def _authors_last_initials(creators: list[str]) -> list[str]:
    """Convert 'Last, First' to 'Last, F.' format."""
    result = []
    for c in creators:
        parts = c.split(",", 1)
        if len(parts) == 2:
            last = parts[0].strip()
            first = parts[1].strip()
            initials = ". ".join(w[0].upper() for w in first.split() if w) + "."
            result.append(f"{last}, {initials}")
        else:
            result.append(c.strip())
    return result


def _authors_apa(creators: list[str]) -> str:
    """Format authors in APA 7th style."""
    formatted = _authors_last_initials(creators)
    if not formatted:
        return "Unknown author"
    if len(formatted) == 1:
        return formatted[0]
    if len(formatted) == 2:
        return f"{formatted[0]} & {formatted[1]}"
    if len(formatted) <= 20:
        return ", ".join(formatted[:-1]) + ", & " + formatted[-1]
    # 20+ authors: first 19, ..., last
    return ", ".join(formatted[:19]) + ", ... " + formatted[-1]


def _authors_harvard(creators: list[str]) -> str:
    """Format authors in Harvard style (same as APA for most cases)."""
    return _authors_apa(creators)


def _authors_chicago(creators: list[str]) -> str:
    """Format authors in Chicago style."""
    formatted = _authors_last_initials(creators)
    if not formatted:
        return "Unknown author"
    if len(formatted) == 1:
        return formatted[0]
    if len(formatted) <= 3:
        return ", ".join(formatted[:-1]) + ", and " + formatted[-1]
    return formatted[0] + " et al."


def _authors_ieee(creators: list[str]) -> list[str]:
    """Format authors in IEEE style (F. Last)."""
    result = []
    for c in creators:
        parts = c.split(",", 1)
        if len(parts) == 2:
            last = parts[0].strip()
            first = parts[1].strip()
            initials = ". ".join(w[0].upper() for w in first.split() if w) + "."
            result.append(f"{initials} {last}")
        else:
            result.append(c.strip())
    return result


def _authors_vancouver(creators: list[str]) -> str:
    """Format authors in Vancouver style."""
    formatted = []
    for c in creators:
        parts = c.split(",", 1)
        if len(parts) == 2:
            last = parts[0].strip()
            first = parts[1].strip()
            initials = "".join(w[0].upper() for w in first.split() if w)
            formatted.append(f"{last} {initials}")
        else:
            formatted.append(c.strip())
    if not formatted:
        return "Unknown author"
    if len(formatted) <= 6:
        return ", ".join(formatted)
    return ", ".join(formatted[:6]) + ", et al"


def _year(record: PrimoRecord) -> str:
    """Extract 4-digit year from creation date."""
    if record.creation_date:
        return record.creation_date[:4]
    return "n.d."


def _publisher_with_place(record: PrimoRecord) -> str:
    """Return "Place: Publisher" when a place is present, else the publisher.

    Used by the styles that retain place of publication (Chicago, IEEE,
    Vancouver); APA 7 and Cite Them Right Harvard (13th ed.) omit place and
    use the plain publisher.
    """
    if record.publisher_place:
        return f"{record.publisher_place}: {record.publisher}"
    return record.publisher


def _terminal_period(value: str) -> str:
    """Append a terminal period unless the string already ends with one."""
    return value if value.endswith(".") else f"{value}."


def _cite_article_apa(r: PrimoRecord) -> str:
    """APA 7 article citation."""
    authors = _authors_apa(r.authors_structured or r.creators)
    year = _year(r)
    title = r.title.rstrip(".")
    journal = r.journal_title

    parts = [f"{authors} ({year}). {title}."]
    if journal:
        vol_info = f"*{journal}*"
        if r.volume:
            vol_info += f", *{r.volume}*"
        if r.issue:
            vol_info += f"({r.issue})"
        if r.start_page:
            vol_info += f", {r.start_page}"
            if r.end_page:
                vol_info += f"-{r.end_page}"
        parts.append(f"{vol_info}.")
    if r.doi:
        parts.append(f"https://doi.org/{r.doi}")
    return " ".join(parts)


def _cite_book_apa(r: PrimoRecord) -> str:
    """APA 7 book citation."""
    authors = _authors_apa(r.authors_structured or r.creators)
    year = _year(r)
    title = r.title.rstrip(".")
    publisher = r.publisher or ""

    parts = [f"{authors} ({year}). *{title}*."]
    if publisher:
        parts.append(_terminal_period(publisher))
    if r.doi:
        parts.append(f"https://doi.org/{r.doi}")
    return " ".join(parts)


def _cite_article_harvard(r: PrimoRecord) -> str:
    """Harvard article citation."""
    authors = _authors_harvard(r.authors_structured or r.creators)
    year = _year(r)
    title = r.title.rstrip(".")
    journal = r.journal_title

    parts = [f"{authors} ({year}) '{title}',"]
    if journal:
        vol_info = f"*{journal}*"
        if r.volume:
            vol_info += f", vol. {r.volume}"
        if r.issue:
            vol_info += f", no. {r.issue}"
        if r.start_page:
            vol_info += f", pp. {r.start_page}"
            if r.end_page:
                vol_info += f"-{r.end_page}"
        parts.append(f"{vol_info}.")
    if r.doi:
        parts.append(f"https://doi.org/{r.doi}")
    return " ".join(parts)


def _cite_book_harvard(r: PrimoRecord) -> str:
    """Harvard book citation."""
    authors = _authors_harvard(r.authors_structured or r.creators)
    year = _year(r)
    title = r.title.rstrip(".")
    publisher = r.publisher or ""

    parts = [f"{authors} ({year}) *{title}*,"]
    if publisher:
        parts.append(_terminal_period(publisher))
    return " ".join(parts)


def _cite_article_chicago(r: PrimoRecord) -> str:
    """Chicago article citation."""
    authors = _authors_chicago(r.authors_structured or r.creators)
    year = _year(r)
    title = r.title.rstrip(".")
    journal = r.journal_title

    parts = [f'{_terminal_period(authors)} "{title}."']
    if journal:
        vol_info = f"*{journal}*"
        if r.volume:
            vol_info += f" {r.volume}"
        if r.issue:
            vol_info += f", no. {r.issue}"
        parts.append(f"{vol_info} ({year})")
        if r.start_page:
            start = r.start_page
            if r.end_page:
                start += f"-{r.end_page}"
            parts[-1] += f": {start}"
        parts[-1] += "."
    if r.doi:
        parts.append(f"https://doi.org/{r.doi}")
    return " ".join(parts)


def _cite_book_chicago(r: PrimoRecord) -> str:
    """Chicago book citation."""
    authors = _authors_chicago(r.authors_structured or r.creators)
    year = _year(r)
    title = r.title.rstrip(".")
    publisher = r.publisher or ""

    parts = [f"{_terminal_period(authors)} *{title}*."]
    if publisher:
        parts.append(f"{_publisher_with_place(r)}, {year}.")
    else:
        parts.append(f"{year}.")
    return " ".join(parts)


def _cite_article_ieee(r: PrimoRecord) -> str:
    """IEEE article citation."""
    authors_list = _authors_ieee(r.authors_structured or r.creators)
    if not authors_list:
        authors_str = "Unknown author"
    elif len(authors_list) <= 3:
        authors_str = ", ".join(authors_list[:-1])
        if len(authors_list) > 1:
            authors_str += ", and " + authors_list[-1]
        else:
            authors_str = authors_list[0]
    else:
        authors_str = authors_list[0] + " et al."

    title = r.title.rstrip(".")
    journal = r.journal_title or ""
    year = _year(r)

    parts = [f'{authors_str}, "{title},"']
    if journal:
        vol_info = f"*{journal}*"
        if r.volume:
            vol_info += f", vol. {r.volume}"
        if r.issue:
            vol_info += f", no. {r.issue}"
        if r.start_page:
            vol_info += f", pp. {r.start_page}"
            if r.end_page:
                vol_info += f"-{r.end_page}"
        parts.append(f"{vol_info}, {year}.")
    if r.doi:
        parts.append(f"doi: {r.doi}.")
    return " ".join(parts)


def _cite_book_ieee(r: PrimoRecord) -> str:
    """IEEE book citation."""
    authors_list = _authors_ieee(r.authors_structured or r.creators)
    if not authors_list:
        authors_str = "Unknown author"
    else:
        authors_str = ", ".join(authors_list[:3])
        if len(authors_list) > 3:
            authors_str += " et al."

    title = r.title.rstrip(".")
    publisher = r.publisher or ""
    year = _year(r)

    parts = [f"{authors_str}, *{title}*."]
    if publisher:
        parts.append(f"{_publisher_with_place(r)}, {year}.")
    else:
        parts.append(f"{year}.")
    return " ".join(parts)


def _cite_article_vancouver(r: PrimoRecord) -> str:
    """Vancouver article citation."""
    authors = _authors_vancouver(r.authors_structured or r.creators)
    title = r.title.rstrip(".")
    journal = r.journal_title or ""
    year = _year(r)

    parts = [f"{authors}. {title}."]
    if journal:
        vol_info = f"{journal}. {year}"
        if r.volume:
            vol_info += f";{r.volume}"
        if r.issue:
            vol_info += f"({r.issue})"
        if r.start_page:
            vol_info += f":{r.start_page}"
            if r.end_page:
                vol_info += f"-{r.end_page}"
        parts.append(f"{vol_info}.")
    if r.doi:
        parts.append(f"doi:{r.doi}")
    return " ".join(parts)


def _cite_book_vancouver(r: PrimoRecord) -> str:
    """Vancouver book citation."""
    authors = _authors_vancouver(r.authors_structured or r.creators)
    title = r.title.rstrip(".")
    publisher = r.publisher or ""
    year = _year(r)

    parts = [f"{authors}. {title}."]
    if publisher:
        parts.append(f"{_publisher_with_place(r)}; {year}.")
    else:
        parts.append(f"{year}.")
    return " ".join(parts)


_STYLE_MAP = {
    "apa7": {"article": _cite_article_apa, "book": _cite_book_apa},
    "harvard": {"article": _cite_article_harvard, "book": _cite_book_harvard},
    "chicago": {"article": _cite_article_chicago, "book": _cite_book_chicago},
    "ieee": {"article": _cite_article_ieee, "book": _cite_book_ieee},
    "vancouver": {"article": _cite_article_vancouver, "book": _cite_book_vancouver},
}


def format_citation(record: PrimoRecord, style: str = "apa7") -> str:
    """Format a citation for a record in the specified style."""
    style_funcs = _STYLE_MAP.get(style, _STYLE_MAP["apa7"])

    # Determine if article or book
    rtype = record.resource_type.lower()
    if rtype in ("article", "review", "newspaper_article"):
        func = style_funcs["article"]
    else:
        func = style_funcs["book"]

    return func(record)

File: src/citations.ts
/**
 * Citation formatting for Primo records (port of citations.py).
 * Supports APA 7th, Harvard, Chicago, IEEE, and Vancouver.
 *
 * Ported from the Python for behavioural parity, with two deliberate
 * deviations: the terminal-period doubling the Python produced (an author
 * initial's trailing dot next to the style's own period, or a publisher ending
 * in "Co.") is de-duplicated via terminalPeriod; and IEEE uses the correct
 * reference-list author rule (all authors up to six, first author + "et al."
 * for seven or more) rather than the Python's three-author cutoff.
 */
import type { PrimoRecord } from "./models.js";

export const CITATION_STYLES = [
  "apa7",
  "harvard",
  "chicago",
  "ieee",
  "vancouver",
] as const;
export type CitationStyle = (typeof CITATION_STYLES)[number];

function authorsFor(record: PrimoRecord): string[] {
  return record.authorsStructured.length > 0
    ? record.authorsStructured
    : record.creators;
}

function stripTrailingDots(value: string): string {
  return value.replace(/\.+$/, "");
}

function yearOf(record: PrimoRecord): string {
  return record.creationDate ? record.creationDate.slice(0, 4) : "n.d.";
}

/**
 * "Place: Publisher" when a place of publication is present, else the
 * publisher alone. Used by the styles that retain place (Chicago, IEEE,
 * Vancouver); APA 7 and Cite Them Right Harvard (13th ed.) omit place and use
 * the plain publisher.
 */
function publisherWithPlace(record: PrimoRecord): string {
  return record.publisherPlace
    ? `${record.publisherPlace}: ${record.publisher}`
    : record.publisher;
}

/** Append a terminal period unless the string already ends with one. */
function terminalPeriod(value: string): string {
  return /\.$/.test(value) ? value : `${value}.`;
}

/** First-name words -> spaced initials with dots, e.g. "Jane Anne" -> "J. A."; "" -> "". */
function initialsDotted(first: string): string {
  const words = first.split(/\s+/).filter((w) => w.length > 0);
  if (words.length === 0) return "";
  return words.map((w) => w.charAt(0).toUpperCase()).join(". ") + ".";
}

/** "Last, First" -> "Last, F. M."; just "Last" when there is no given name; unchanged if there is no comma. */
function authorsLastInitials(creators: string[]): string[] {
  return creators.map((c) => {
    const idx = c.indexOf(",");
    if (idx === -1) return c.trim();
    const last = c.slice(0, idx).trim();
    const first = c.slice(idx + 1).trim();
    const initials = initialsDotted(first);
    return initials ? `${last}, ${initials}` : last;
  });
}

function authorsApa(creators: string[]): string {
  const f = authorsLastInitials(creators);
  if (f.length === 0) return "Unknown author";
  if (f.length === 1) return f[0] ?? "";
  if (f.length === 2) return `${f[0]} & ${f[1]}`;
  if (f.length <= 20) return f.slice(0, -1).join(", ") + ", & " + f[f.length - 1];
  return f.slice(0, 19).join(", ") + ", ... " + f[f.length - 1];
}

function authorsChicago(creators: string[]): string {
  const f = authorsLastInitials(creators);
  if (f.length === 0) return "Unknown author";
  if (f.length === 1) return f[0] ?? "";
  if (f.length <= 3) return f.slice(0, -1).join(", ") + ", and " + f[f.length - 1];
  return f[0] + " et al.";
}

/** IEEE author list entries: "F. M. Last"; just "Last" when there is no given name. */
function authorsIeee(creators: string[]): string[] {
  return creators.map((c) => {
    const idx = c.indexOf(",");
    if (idx === -1) return c.trim();
    const last = c.slice(0, idx).trim();
    const first = c.slice(idx + 1).trim();
    const initials = initialsDotted(first);
    return initials ? `${initials} ${last}` : last;
  });
}

/**
 * IEEE reference-list author string. IEEE lists all authors up to six; for
 * seven or more it gives the first author followed by "et al." (no comma before
 * et al.). Two authors are joined with "and" (no serial comma); three to six use
 * a serial comma before the final "and".
 */
function authorsIeeeFormatted(creators: string[]): string {
  const list = authorsIeee(creators);
  if (list.length === 0) return "Unknown author";
  if (list.length === 1) return list[0] ?? "";
  if (list.length === 2) return `${list[0]} and ${list[1]}`;
  if (list.length <= 6) {
    return list.slice(0, -1).join(", ") + ", and " + list[list.length - 1];
  }
  return `${list[0]} et al.`;
}

function authorsVancouver(creators: string[]): string {
  const f = creators.map((c) => {
    const idx = c.indexOf(",");
    if (idx === -1) return c.trim();
    const last = c.slice(0, idx).trim();
    const first = c.slice(idx + 1).trim();
    const initials = first
      .split(/\s+/)
      .filter((w) => w.length > 0)
      .map((w) => w.charAt(0).toUpperCase())
      .join("");
    return initials ? `${last} ${initials}` : last;
  });
  if (f.length === 0) return "Unknown author";
  if (f.length <= 6) return f.join(", ");
  return f.slice(0, 6).join(", ") + ", et al";
}

function citeArticleApa(r: PrimoRecord): string {
  const authors = authorsApa(authorsFor(r));
  const year = yearOf(r);
  const title = stripTrailingDots(r.title);
  const parts: string[] = [`${authors} (${year}). ${title}.`];
  if (r.journalTitle) {
    let volInfo = `*${r.journalTitle}*`;
    if (r.volume) volInfo += `, *${r.volume}*`;
    if (r.issue) volInfo += `(${r.issue})`;
    if (r.startPage) {
      volInfo += `, ${r.startPage}`;
      if (r.endPage) volInfo += `-${r.endPage}`;
    }
    parts.push(`${volInfo}.`);
  }
  if (r.doi) parts.push(`https://doi.org/${r.doi}`);
  return parts.join(" ");
}

function citeBookApa(r: PrimoRecord): string {
  const authors = authorsApa(authorsFor(r));
  const year = yearOf(r);
  const title = stripTrailingDots(r.title);
  const parts: string[] = [`${authors} (${year}). *${title}*.`];
  if (r.publisher) parts.push(terminalPeriod(r.publisher));
  if (r.doi) parts.push(`https://doi.org/${r.doi}`);
  return parts.join(" ");
}

function citeArticleHarvard(r: PrimoRecord): string {
  const authors = authorsApa(authorsFor(r));
  const year = yearOf(r);
  const title = stripTrailingDots(r.title);
  const parts: string[] = [`${authors} (${year}) '${title}',`];
  if (r.journalTitle) {
    let volInfo = `*${r.journalTitle}*`;
    if (r.volume) volInfo += `, vol. ${r.volume}`;
    if (r.issue) volInfo += `, no. ${r.issue}`;
    if (r.startPage) {
      volInfo += `, pp. ${r.startPage}`;
      if (r.endPage) volInfo += `-${r.endPage}`;
    }
    parts.push(`${volInfo}.`);
  }
  if (r.doi) parts.push(`https://doi.org/${r.doi}`);
  return parts.join(" ");
}

function citeBookHarvard(r: PrimoRecord): string {
  const authors = authorsApa(authorsFor(r));
  const year = yearOf(r);
  const title = stripTrailingDots(r.title);
  const parts: string[] = [`${authors} (${year}) *${title}*,`];
  if (r.publisher) parts.push(terminalPeriod(r.publisher));
  return parts.join(" ");
}

function citeArticleChicago(r: PrimoRecord): string {
  const authors = authorsChicago(authorsFor(r));
  const year = yearOf(r);
  const title = stripTrailingDots(r.title);
  const parts: string[] = [`${terminalPeriod(authors)} "${title}."`];
  if (r.journalTitle) {
    let volInfo = `*${r.journalTitle}*`;
    if (r.volume) volInfo += ` ${r.volume}`;
    if (r.issue) volInfo += `, no. ${r.issue}`;
    let segment = `${volInfo} (${year})`;
    if (r.startPage) {
      let pages = r.startPage;
      if (r.endPage) pages += `-${r.endPage}`;
      segment += `: ${pages}`;
    }
    segment += ".";
    parts.push(segment);
  }
  if (r.doi) parts.push(`https://doi.org/${r.doi}`);
  return parts.join(" ");
}

function citeBookChicago(r: PrimoRecord): string {
  const authors = authorsChicago(authorsFor(r));
  const year = yearOf(r);
  const title = stripTrailingDots(r.title);
  const parts: string[] = [`${terminalPeriod(authors)} *${title}*.`];
  if (r.publisher) parts.push(`${publisherWithPlace(r)}, ${year}.`);
  else parts.push(`${year}.`);
  return parts.join(" ");
}

function citeArticleIeee(r: PrimoRecord): string {
  const authors = authorsIeeeFormatted(authorsFor(r));
  const title = stripTrailingDots(r.title);
  const year = yearOf(r);
  const parts: string[] = [`${authors}, "${title},"`];
  if (r.journalTitle) {
    let volInfo = `*${r.journalTitle}*`;
    if (r.volume) volInfo += `, vol. ${r.volume}`;
    if (r.issue) volInfo += `, no. ${r.issue}`;
    if (r.startPage) {
      volInfo += `, pp. ${r.startPage}`;
      if (r.endPage) volInfo += `-${r.endPage}`;
    }
    parts.push(`${volInfo}, ${year}.`);
  }
  if (r.doi) parts.push(`doi: ${r.doi}.`);
  return parts.join(" ");
}

function citeBookIeee(r: PrimoRecord): string {
  const authors = authorsIeeeFormatted(authorsFor(r));
  const title = stripTrailingDots(r.title);
  const year = yearOf(r);
  const parts: string[] = [`${authors}, *${title}*.`];
  if (r.publisher) parts.push(`${publisherWithPlace(r)}, ${year}.`);
  else parts.push(`${year}.`);
  return parts.join(" ");
}

function citeArticleVancouver(r: PrimoRecord): string {
  const authors = authorsVancouver(authorsFor(r));
  const title = stripTrailingDots(r.title);
  const year = yearOf(r);
  const parts: string[] = [`${authors}. ${title}.`];
  if (r.journalTitle) {
    let volInfo = `${r.journalTitle}. ${year}`;
    if (r.volume) volInfo += `;${r.volume}`;
    if (r.issue) volInfo += `(${r.issue})`;
    if (r.startPage) {
      volInfo += `:${r.startPage}`;
      if (r.endPage) volInfo += `-${r.endPage}`;
    }
    parts.push(`${volInfo}.`);
  }
  if (r.doi) parts.push(`doi:${r.doi}`);
  return parts.join(" ");
}

function citeBookVancouver(r: PrimoRecord): string {
  const authors = authorsVancouver(authorsFor(r));
  const title = stripTrailingDots(r.title);
  const year = yearOf(r);
  const parts: string[] = [`${authors}. ${title}.`];
  if (r.publisher) parts.push(`${publisherWithPlace(r)}; ${year}.`);
  else parts.push(`${year}.`);
  return parts.join(" ");
}

type StyleFuncs = {
  article: (r: PrimoRecord) => string;
  book: (r: PrimoRecord) => string;
};

const STYLE_MAP: Record<string, StyleFuncs> = {
  apa7: { article: citeArticleApa, book: citeBookApa },
  harvard: { article: citeArticleHarvard, book: citeBookHarvard },
  chicago: { article: citeArticleChicago, book: citeBookChicago },
  ieee: { article: citeArticleIeee, book: citeBookIeee },
  vancouver: { article: citeArticleVancouver, book: citeBookVancouver },
};

const ARTICLE_TYPES = new Set(["article", "review", "newspaper_article"]);

export function formatCitation(record: PrimoRecord, style = "apa7"): string {
  const styleFuncs = STYLE_MAP[style] ?? { article: citeArticleApa, book: citeBookApa };
  const rtype = record.resourceType.toLowerCase();
  const func = ARTICLE_TYPES.has(rtype) ? styleFuncs.article : styleFuncs.book;
  return func(record);
}
